File: LeetCode/76MinimumWindowSubString/MinimumWindowSubString.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # Single Pass -> Time:O(n), Space:O(n)
        t_d = {}
        sub_str = {}
        for c in t:
            t_d[c] = 1 + t_d.get(c, 0)
        have, need = 0, len(t_d)
        l = 0
        res, res_len = [-1, -1], len(s)+1
        for r in range(len(s)):
            c = s[r]
            sub_str[c] = 1 + sub_str.get(c, 0)
            if c in t_d and sub_str[c] == t_d[c]:
                have += 1
            while have == need:
                if r-l+1 < res_len:
                    res = [l, r]
                    res_len = r-l+1
                sub_str[s[l]] -= 1
                if s[l] in t_d and sub_str[s[l]] < t_d[s[l]]:
                    have -= 1
                l += 1
        return s[res[0]:res[1]+1] if res_len != len(s)+1 else ""

        # # Brute Force -> Time:O(n^2), Space:O(n)
        # if len(t) > len(s):
        #     return ""
        # t_d = {}
        # for c in t:
        #     if c not in t_d:
        #         t_d[c] = 0
        #     t_d[c] += 1
        # for i in range(len(t), len(s)+1):
        #     for j in range(len(s)-i+1):
        #         sub_str_d = {}
        #         sub_str = s[j:j+i]
        #         for c in sub_str:
        #             if c not in sub_str_d:
        #                 sub_str_d[c] = 0
        #             sub_str_d[c] += 1
        #         min_window = True
        #         for c in t_d:
        #             if c not in sub_str_d:
        #                 min_window = False
        #             elif sub_str_d[c] < t_d[c]:
        #                 min_window = False
        #         if min_window:
        #             return sub_str
        # return ""

File: LeetCode/76MinimumWindowSubString/MinimumWindowSubString.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        // Single Pass -> Time:O(n), Space:O(n)
        unordered_map<char, int> tmap;
        for (int i = 0; i < t.size(); i++) {
            tmap[t[i]]++;
        }
        int l = 0, have = 0, need = tmap.size();
        vector<int> res {-1, -1};
        int res_len = s.size()+1;
        char c;
        unordered_map<char, int> sub_str_map;
        for (int r = 0; r < s.size(); r++) {
            c = s[r];
            sub_str_map[c]++;
            if (tmap.find(c) != tmap.end() && tmap[c] == sub_str_map[c]) {
                have++;
            }
            while (have == need) {
                if (r-l+1 < res_len) {
                    res[0] = l;
                    res[1] = r;
                    res_len = r-l+1;
                }
                sub_str_map[s[l]] -= 1;
                if (tmap.find(s[l]) != tmap.end() && sub_str_map[s[l]] < tmap[s[l]]) {
                    have--;
                }
                l++;
            }
        }
        if (res_len == s.size()+1) {
            return "";
        }
        return s.substr(res[0], res[1]-res[0]+1);
    }
};
